tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Session.java

```diff
@@ -225,24 +235,59 @@ public Runner fetch(String operation) {
      */
     public Runner fetch(String operation, int index) {
       Operation op = graph.operationOrThrow(operation);
-      outputs.add(op.output(index));
-      return this;
+      return fetch(op.output(index));
     }
 
     /**
      * Makes {@link #run()} return the Tensor referred to by {@code output}.
      *
+     * If {@code output} is a resource variable, will fetch the value.
+     *
      * @param output the node to fetch the tensor from
      * @return this session runner
      */
     public Runner fetch(Output<?> output) {
-      outputs.add(output);
+      if (output.env() != graph) {
+        throw new IllegalStateException("Can't fetch output " + output + ", it is from a different graph.");
+      }
+
+      if (output.dataType() == DataType.DT_RESOURCE) {
+        int[] rawDt = new int[1];
+
+        GraphOperation graphOp = (GraphOperation) output.op();
+
+        try (PointerScope scope = new PointerScope()) {
+          TF_Status status = TF_Status.newStatus();
+          TF_OperationGetAttrType(graphOp.getUnsafeNativeHandle(), "dtype", rawDt, status);
+          status.throwExceptionIfNotOK();
+        }
+
+        DataType valueDt = DataType.forNumber(rawDt[0]);
+
+        Operand<?> read = null;
+        for (GraphOperation op : graphOp.consumers()) {
+          if (op.dtype(0) == valueDt && op.type().equals(ReadVariableOp.OP_NAME)) {
+            read = op.output(0);
+          }
+        }
+
+        if (read == null) {
+          read = Ops.create(graph).withSubScope("session_reads").withName(output.op().name() + "_read")
```

Comment:
Might be nice to have some logging here that we can turn on to see if it's being hit a lot (as that means our assumptions about graphs are being violated). Maybe at `logger.debug` level?

Reply:
I don't think we have a logging library at the moment, do we?  But yeah, it would be nice.

Reply:
Oh yeah, probably not. I tend to use `java.util.logging` for everything but I understand that most people use `slf4j`.

Reply:
There's been a couple times that having one would be nice, we should decide on something.  Next core meeting maybe?

Reply:
Yeah, let's discuss it then.